File: Algorithms/quick Sort/QuickSort.java
public class QuickSort {
    public static void main(String args[]) {
        int[] arr = { 12, 11, 13, 5, 6, 7, 32523532, 13 };

        System.out.println("Given Array");
        printArray(arr);

        quickSort(arr, 0, arr.length - 1);

        System.out.println("\nSorted array");
        printArray(arr);
    }

    public static void printArray(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void quickSort(int[] arr, int l, int r) {
        if (l < r) {
            int j = partition(arr, l, r);
                quickSort(arr, j + 1, r);
                quickSort(arr, l, j - 1);
        }
    }

    public static int partition(int[] arr, int l, int r) {
        int p = arr[l];
        int i = l, j = r;

        while (i < j) {
            while (arr[i] <= p && i < r)  // here should be arr[i] <= p, because j is swapping 
                i++;
            while (arr[j] > p && j >= l)
                j--;
            if (i < j) {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[l];
        arr[l] = arr[j];
        arr[j] = temp;

        return j;
    }

}

File: Algorithms/quick Sort/quicksort.cpp
#include <iostream>
#include <algorithm>

using namespace std;

void printArray(int arr[], int size)
{
    for (int i = 0; i < size; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

size_t partition(int *arr, size_t l, size_t r)
{
    int p = arr[l]; // Pivot
    size_t i = l, j = r;

    while (i < j)
    {
        while (i <= r && arr[i] <= p) i++; // Find element > pivot
        while (j >= l && arr[j] > p) j--;  // Find element <= pivot

        if (i < j)
            swap(arr[i], arr[j]); // Swap misplaced elements
    }

    swap(arr[l], arr[j]); // Place pivot at correct position
    return j; // Return pivot index
}

void quickSort(int *arr, size_t l, size_t r)
{
    if (l < r)
    {
        size_t j = partition(arr, l, r);
        if (j > l) quickSort(arr, l, j - 1);
        if (j + 1 < r) quickSort(arr, j + 1, r);
    }
}

int main()
{
    int arr[] = {12, 11, 13, 5, 6, 1};
    int size = sizeof(arr) / sizeof(arr[0]);

    cout << "Given array: ";
    printArray(arr, size);

    quickSort(arr, 0, size - 1);

    cout << "Sorted array: ";
    printArray(arr, size);

    return 0;
}